var/spack/repos/builtin/packages/gcc/package.py

```diff
@@ -503,11 +501,22 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
 
     build_directory = "spack-build"
 
-    @classproperty
-    def executables(cls):
-        names = [r"gcc", r"[^\w]?g\+\+", r"gfortran", r"gdc", r"gccgo"]
-        suffixes = [r"", r"-mp-\d+\.\d", r"-\d+\.\d", r"-\d+", r"\d\d"]
-        return [r"".join(x) for x in itertools.product(names, suffixes)]
+    compiler_languages = ["c", "cxx", "fortran", "d", "go"]
+
+    @property
+    def supported_languages(self):
+        # This weirdness is because it could be called on an abstract spec
+        if "languages" not in self.spec.variants:
+            return self.compiler_languages
+        return [x for x in self.compiler_languages if x in self.spec.variants["languages"].value]
+
+    c_names = ["gcc"]
+    cxx_names = ["g++"]
+    fortran_names = ["gfortran"]
+    d_names = ["gdc"]
+    go_names = ["gccgo"]
+    prefixes = [r"\w+-\w+-\w+-"]
+    suffixes = [r"-mp-\d+(?:\.\d+)?", r"-\d+(?:\.\d+)?", r"\d\d"]
```

Comment:
Here we changed the suffixes and prefixes. Why?

Reply:
The first suffix in the old code is included automatically by the `CompilerPackage` class.

The second suffix in the old code is covered by the first suffix here, but the regex is made more general to match the behavior from the `spack.compilers.gcc.Gcc`. 

The third and fourth suffixes in the old code, when combined, are identical to the second suffix here.

The fifth suffix in the old code is the fourth suffix here.

The prefixes here are added to better match the behavior of `spack.compilers.gcc.Gcc`.

Reply:
> The first suffix in the old code is included automatically by the CompilerPackage class.

I'm not particularly fond of that, but let's postpone the discussion to another PR.